core/src/main/java/io/grpc/internal/AutoConfiguredLoadBalancerFactory.java

```diff
@@ -155,73 +169,97 @@ void setDelegate(LoadBalancer lb) {
-      return new PolicySelection(provider, null);
-    }
 
-    if (config != null) {
-      List<Map<String, Object>> lbConfigs =
-          ServiceConfigUtil.getLoadBalancingConfigsFromServiceConfig(config);
-      LinkedHashSet<String> policiesTried = new LinkedHashSet<>();
-      for (Map<String, Object> lbConfig : lbConfigs) {
-        if (lbConfig.size() != 1) {
-          throw new PolicyException(
-              "There are " + lbConfig.size()
-              + " load-balancing configs in a list item. Exactly one is expected. Config="
-              + lbConfig);
+      if (haveBalancerAddress) {
+        LoadBalancerProvider grpclbProvider = registry.getProvider(GRPCLB_POLICY);
+        if (grpclbProvider == null) {
+          if (backendAddrs.isEmpty()) {
+            throw new PolicyException(
+                "Received ONLY balancer addresses but grpclb runtime is missing");
+          }
+          if (!roundRobinDueToGrpclbDepMissing) {
+            roundRobinDueToGrpclbDepMissing = true;
+            String errorMsg = "Found balancer addresses but grpclb runtime is missing."
+                + " Will use round_robin. Please include grpc-grpclb in your runtime depedencies.";
+            helper.getChannelLogger().log(ChannelLogLevel.ERROR, errorMsg);
+            logger.warning(errorMsg);
+          }
+          return new PolicySelection(
+              getProviderOrThrow(
+                  "round_robin", "received balancer addresses but grpclb runtime is missing"),
+              backendAddrs, null);
+        } else {
+          return new PolicySelection(grpclbProvider, servers, null);
         }
-        Entry<String, Object> entry = lbConfig.entrySet().iterator().next();
-        String policy = entry.getKey();
-        LoadBalancerProvider provider = registry.getProvider(policy);
-        if (provider != null) {
-          if (!policiesTried.isEmpty()) {
-            channelLogger.log(
-                ChannelLogLevel.DEBUG,
-                "{0} specified by Service Config are not available", policiesTried);
+      }
+      roundRobinDueToGrpclbDepMissing = false;
+
+      if (config != null) {
+        List<Map<String, Object>> lbConfigs =
+            ServiceConfigUtil.getLoadBalancingConfigsFromServiceConfig(config);
+        LinkedHashSet<String> policiesTried = new LinkedHashSet<>();
+        for (Map<String, Object> lbConfig : lbConfigs) {
+          if (lbConfig.size() != 1) {
+            throw new PolicyException(
+                "There are " + lbConfig.size()
+                + " load-balancing configs in a list item. Exactly one is expected. Config="
+                + lbConfig);
+          }
+          Entry<String, Object> entry = lbConfig.entrySet().iterator().next();
+          String policy = entry.getKey();
+          if (policy.equals(GRPCLB_POLICY)) {
+            throw new PolicyException(
+                GRPCLB_POLICY + " specified by Service Config but no balancer addresses exist");
```

Comment:
This looks like it should exist in gRPC LB instead. And it seems like something like that is already in GrpclbState.

Reply:
Done.